src/mlpack/methods/reinforcement_learning/replay/sumtree.hpp

```diff
@@ -0,0 +1,194 @@
+/**
+ * @file sumtree.hpp
+ * @author Xiaohong
+ *
+ * This file is an implementation of sumtree.
+ *
+ * reference:
+ * [1] https://github.com/openai/baselines/blob/master/baselines/common/segment_tree.py
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_RL_SUMTREE_HPP
+#define MLPACK_METHODS_RL_SUMTREE_HPP
+
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack {
+namespace rl {
+
+/**
+ * Implementation of SumTree.
+ *
+ * Build a Segment Tree like data structure.
+ * https://en.wikipedia.org/wiki/Segment_tree
+ *
+ * Used to maintain prefix-sum of an array.
+ *
+ */
+
+template<typename T>
+class SumTree
```

Comment:
I wonder if it makes sense to implement the Segment Tree following: https://github.com/mlpack/mlpack/wiki/OldSummerOfCodeIdeas#implement-tree-types let me know what you think. @rcurtin might be able to provide some more insight.

Reply:
I remember discussing the segment tree sometime years back, and my conclusion was that it wouldn't really fit in to the `TreeType` abstraction since it's not a multidimensional tree.  Another problem might be (I think) that the segment tree holds segments, not individual points, so it could be a little bit hard to fit that in to the current abstraction.  Maybe there are some ways it could be made to fit, but since segment trees (or sum trees) aren't used directly for distance-like computations in the same ways the existing trees are, I don't know how much overlap there is.

I'm not saying it's not possible of course, just that it might be too tricky and not worth it here.  I don't see a problem with having a separate `SumTree` structure if it's needed (assuming it doesn't already exist in some form in the STL or something). :)